contracts/test/v0.8/dev/UpkeepTranscoder3_0.test.ts

```diff
@@ -0,0 +1,216 @@
+import { ethers } from 'hardhat'
+import { assert } from 'chai'
+import { UpkeepTranscoder30__factory as UpkeepTranscoderFactory } from '../../../typechain/factories/UpkeepTranscoder30__factory'
+import { UpkeepTranscoder30 as UpkeepTranscoder } from '../../../typechain/UpkeepTranscoder30'
+import { evmRevert } from '../../test-helpers/matchers'
+
+let upkeepTranscoderFactory: UpkeepTranscoderFactory
+let transcoder: UpkeepTranscoder
+let balance: number
+let executeGas: number
+let amountSpent: number
+let admin0: string
+let target0: string
+let lastKeeper0: string
+let admin1: string
+let target1: string
+let lastKeeper1: string
+let idx: number[]
+let upkeepsV1: any[]
+let upkeepsV2: any[]
+let upkeepsV3: any[]
+let admins: string[]
+
+const encodeUpkeepV1 = (ids: number[], upkeeps: any[], checkDatas: any[]) => {
+  return ethers.utils.defaultAbiCoder.encode(
+    [
+      'uint256[]',
+      'tuple(uint96,address,uint32,uint64,address,uint96,address)[]',
+      'bytes[]',
+    ],
+    [ids, upkeeps, checkDatas],
+  )
+}
+
+const encodeUpkeepV2 = (ids: number[], upkeeps: any[], checkDatas: any[]) => {
+  return ethers.utils.defaultAbiCoder.encode(
+    [
+      'uint256[]',
+      'tuple(uint96,address,uint96,address,uint32,uint32,address,bool)[]',
+      'bytes[]',
+    ],
+    [ids, upkeeps, checkDatas],
+  )
+}
+
+const encodeUpkeepV3 = (
+  ids: number[],
+  upkeeps: any[],
+  checkDatas: any[],
+  admins: string[],
+) => {
+  return ethers.utils.defaultAbiCoder.encode(
+    [
+      'uint256[]',
+      'tuple(uint32,uint32,bool,address,uint96,uint96,uint32)[]',
+      'bytes[]',
+      'address[]',
+    ],
+    [ids, upkeeps, checkDatas, admins],
+  )
+}
+
+before(async () => {
+  // @ts-ignore bug in autogen file
+  upkeepTranscoderFactory = await ethers.getContractFactory(
+    'UpkeepTranscoder3_0',
+  )
+})
+
+describe.only('UpkeepTranscoder3_0', () => {
```

Comment:
could you write a test which actually deploys the registries of different version and calls migrate.

Would test the e2e flow and much better coverage

Reply:
Done